File: leetcode/202-happy-number/solution.py
class Solution:
    def sum_of_squares(self, n: int) -> int:
        total = 0
        while n > 0:
            d = n % 10
            total += d * d
            n //= 10
        return total

    def isHappy(self, n: int) -> bool:
        seen = set()
        while n != 1 and n not in seen:
            seen.add(n)
            n = self.sum_of_squares(n)
        return n == 1


if __name__ == "__main__":
    s = Solution()
    print(s.isHappy(19))  # True
    print(s.isHappy(2))  # False

File: leetcode/202-happy-number/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int sumOfSquares(int n) {
        int sum = 0;
        while (n > 0) {
            int d = n % 10;
            sum += d * d;
            n /= 10;
        }
        return sum;
    }

    bool isHappy(int n) {
        unordered_set<int> seen;
        while (n != 1 && !seen.count(n)) {
            seen.insert(n);
            n = sumOfSquares(n);
        }
        return n == 1;
    }
};

int main() {
    Solution sol;
    cout << sol.isHappy(19) << endl; // 1 (true)
    cout << sol.isHappy(2) << endl;  // 0 (false)
    return 0;
}
